design/user-profile.md

```diff
@@ -206,9 +250,16 @@ All mappers that use user attributes should provide a drop-down with list of att
 
 ## Milestones
 
-### M1 SPI
+### M1 SPI, Legacy Provider and UserModel preparation/refactoring
+
+Define the SPI/API for the user profile, its attributes with validators (and optionally permissions).
+Implement simple legacy provider.
+
+Migrate to new user model where all fields like username, email, etc. are attributes
+Don't change database model: Use entity - model adapter classes to enforce legacy behaviour: all legacy fields are "mandatory"
```

Comment:
Do you mean removing getFirstName/LastName/Email fields on UserModel?

Reply:
yes, we would remove it completely in the whole middle tier. It will only stay in the representation/context and the entity in the first revision we plan right now. The idea is to partially fade away from the notion of explicit fields of a user. Instead, there are some well-known attribute keys e.g. username. 

Our thought here is that in the future there will be no "column" username as there are no columns.. so specific handling in the model for this field seems not necessary. 

This would enable consistent handling of the profile and its values in the whole application as there are no static fields anymore. And we can define (hardcoded) validators for all fields also the username and email. 

Do you have concerns that the username should stay a fixed field? 
We could easily agree that e.g. username and email stay fixed as they have special meanings regards identity of a user. 
The question is would you like to be able to have validation functions "non hardcoded" on those fields. if not we can leave them static. But it would be hard to attach validators to nondynamic fields. 
But on the other hand, we would like to have mostly dynamic handling of user information.

Reply:
I agree that most things should be attributes, not fields. Even the JSON representation at some point needs a cleanup, but that is a separate thing to this. There's quite some things that needs to be figured out though to be able to update the UserModel though.

Keycloak uses username, email and email-verified quite a lot, and needs a fixed way to access this. Further, username and email use DB constraints to make sure they are unique. Not having them as columns complicates that. Then again, emails sometimes should be unique and sometimes not. User sometimes have a username, and sometimes only an email to login with.

Reply:
Yes, that's something we are also aware of (and therefore we wanted to specifically address this here). Our idea was the following:

We don't want to change the database model - at least in the beginning, so the columns would stay the same (in other words: the UserEntity wouldn't be touched). We would make sure that the UserAdapter and the RealmAdapter ensure that the user fields are correctly set and handled. Then you could figure out later what to do with the database model, whether or not to remove the columns (I guess users may have indices on the username column or the like so that needs to be very carefully considered), etc.

We do however want to also switch the fields to attributes in the UserRepresentation and therefore the JSON representation, but we think this is best done later when working on the UI to reflect the profile, because it seems to me that it is more difficult to ensure this is done correctly and we'd need more familiarity with the codebase in the UI first. 

We see this as a tradeoff: 
- When modelling fields as attributes in the UserModel (and representation), this would allow a simpler and cleaner implementation of the user profile and at the same time allow us to cleanly enforce things like unique usernames and/or emails and make that behaviour configurable in the profile later on
- Incremental changes of the model leave us with an awkward intermediate state where e.g. the UserModel class doesn't know about "First name" while the representation and entity classes still do, but then smaller changes are easier to get right (and review)
- From what we gather in your design documents, you wish to migrate to this kind of user model where many things are attributes anyway, so we thought we could maybe make a step in this direction now but keeping the old model in the database

So in order to switch the model, here is how I would proceed for e.g. the last name (email and username would be slightly more difficult probably):
- Define a class like `BasicUserAttributes` or maybe `OIDCBasicClaimAttributes` or similar where we define a constant `GIVEN_NAME="given_name"` as the last name could match the corresponding open id claim.
- Replace all occurrences of `UserModel.setLastName(name)` by `UserModel.setSingleAttribute(GIVEN_NAME, name)` (and similar for the getter). 
- In the database adapter (UserAdapter) we'd rewrite the methods setting and getting attributes to add the given name to the attributes (getting attributes) or to call `entity.setLastName` if the attribute to set was `GIVEN_NAME`. That way, the last name attribute should still be saved and read from the corresponding column, but internally, it's being added to the attributes and used as if the field didn't exist. There is of course some runtime overhead to this, but I'd hope it's small enough to not matter.
- Similarly, the `RepresentationToModel` classes (and vice versa) have to map between the field in the representation and the attribute in the model until the field gets removed from the representation later on.

If you think we are missing something or would like to see an even more concrete plan, I can open a PR implementing these changes to further discuss this issue.

Reply:
In general this sounds like a sensible plan. Probably worth dropping an email around this on the dev mailing list separately and have this done in a PR as a pre-requisite to this work.

Reply:
The plan would be to do this in a separate PR, it sounds like a good idea to announce this in a separate mailing list thread - I'll do that soon.